### PR TITLE
36 connecting to s3 - Mohamed Abdalla

### LINKED ISSUE #36: Connecting to S3
We are using AWS S3 to store images. following this steps:

- Install Multer, multer-s3 and aws sdk
- Setup the S3 client
- Create our uploadImage middleware using multer
- Use uploadImage middleware in routes

### PR DESCRIPTION
**Pull request by Mohamed-Abdalla-1**
## What?

This PR is to connect to AWS S3 storage in where we save the uploaded images.
Used multer and AWS packages and updated the dummy upload API.

## Why?

This will enable us to us the upload API to store images in S3 storage.

## Related to

Connecting to S3 #36

### CODE REVIEW

File: server/api.js [hunk 1]
```diff
@@ -27,8 +24,27 @@ router.get("/image", (req, res) => {
...
+router.put(
+	"/image",
+	uploadImage.single("image"), // our uploadImage middleware
+	(req, res) => {
+		/*
+           req.file = {
+             fieldname, originalname,
+             mimetype, size, bucket, key, location
+           }
+        */
```
Comment (oliverlloyd) [inline server/api.js]: We could delete this commented out code

File: server/api.js [hunk 2]
```diff
@@ -27,8 +24,27 @@ router.get("/image", (req, res) => {
...
+		console.log("request", req.file);
```
Comment (oliverlloyd) [inline server/api.js]: We should remove console log statements from code we push. If we need logs, we can use the `logger` feature

File: server/utils/uploadImage.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+const s3 = new S3Client({
+	credentials: {
+		accessKeyId: process.env.ACCESS_KEY_ID, // store it in .env file to keep it safe
+		secretAccessKey: process.env.SECRET_ACCESS_KEY,
+	},
+	region: "eu-west-1", // this is the region that you select in AWS account
+});
+
+const s3Storage = multerS3({
+	s3: s3, // s3 instance
```
Comment (oliverlloyd) [inline server/utils/uploadImage.js]: Minor nitpick: If the variable name is the same, you can write it like this
```suggestion
	s3, // s3 instance
```

File: server/utils/uploadImage.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	);
+
+	// Mime type must be an image
+	const isAllowedMimeType = file.mimetype.startsWith("image/");
+
+	if (isAllowedExt && isAllowedMimeType) {
+		return cb(null, true); // no errors
+	} else {
+		// pass error msg to callback, which can be displayed in frontend
+		cb("Error: File type not allowed!");
```
Comment (oliverlloyd) [inline server/utils/uploadImage.js]: We haven't discussed error handling but ideally error messages would be show to a user on the frontend so they understand what went wrong and can fix the problem and try again

File: server/exampleData.json [hunk 5]
```diff
@@ -2,19 +2,20 @@
 	{
 		"id": 523523,
 		"description": "Description",
-		"src":"https://i.ibb.co/1TB3Kmv/CYF-Images-bank-gallery.jpg",
+		"src": "https://cyf-images-fp.s3.eu-west-1.amazonaws.com/1687617876279_image_IMG_4743.jpg",
```
Comment (oliverlloyd) [inline server/exampleData.json:5]: Great to see our first example image 🚀

**Review by oliverlloyd — APPROVED** [review]
This is a big step!

I've made some small comments but I'm happy for you to address these and merge when you're ready 👍

File: server/api.js (hunk 1 above)
Comment (Mohamed-Abdalla-1) [reply server/api.js]: Yes, thanks. I fixed in this commit: https://github.com/Innapoliakova/set/pull/50/commits/53b418d3a4cc031d76573d7a38a100c480404520

File: server/api.js (hunk 2 above)
Comment (Mohamed-Abdalla-1) [reply server/api.js]: Yes, I have removed it for now in this commit: https://github.com/Innapoliakova/set/pull/50/commits/53b418d3a4cc031d76573d7a38a100c480404520

File: server/utils/uploadImage.js (hunk 4 above)
Comment (Mohamed-Abdalla-1) [reply server/utils/uploadImage.js]: I've altered the message to be clearer in this commit:https://github.com/Innapoliakova/set/pull/50/commits/22ce2924421ae0cb032dc963d1a4d26a1c084de4

Comment (oliverlloyd) [reply server/utils/uploadImage.js]: Nice!

**Pull request merged**